src/lua/uri.c

```diff
@@ -0,0 +1,390 @@
+
+/**
+ * Check if there is a field with the name @name in the table,
+ * which located which located at the given valid @a idx.
+ */
+static bool
+is_field_present(struct lua_State *L, const char *name, int idx)
+{
+	assert(lua_type(L, idx) == LUA_TTABLE);
+	lua_pushstring(L, name);
+	lua_rawget(L, idx);
+	bool field_is_present = (lua_type(L, -1) != LUA_TNIL);
+	lua_pop(L, 1);
+	return field_is_present;
+}
+
+/**
+ * Create @ uri from the string or number value at the top of lua
+ * stack. Return -1 if value has invalid type.
+ */
+static int
+uri_create_from_string_or_number(struct uri *uri, struct lua_State *L)
+{
+	const char *errmsg;
+	int rc = 0;
+	switch (lua_type(L, -1)) {
+	case LUA_TSTRING:
+	case LUA_TNUMBER:
+		rc = uri_create(uri, lua_tostring(L, -1));
+		if (rc != 0) {
+			errmsg = "Incorrect URI: expected host:service or "
+				 "/unix.socket";
+			diag_set(IllegalParams, errmsg);
+		}
+		break;
+	default:
+		errmsg = "Incorrect type for URI in nested table: "
+			 "should be string, number";
+		diag_set(IllegalParams, errmsg);
+		rc = -1;
+	}
+	return rc;
+}
+
+/**
+ * Create @a uri from the table, which located at the given valid @a idx.
+ */
+static int
+uri_create_from_lua_table(struct uri *uri, struct lua_State *L, int idx)
+{
+	assert(lua_type(L, idx) == LUA_TTABLE);
+	/* There must be exactly one URI in the table */
+	int size = lua_objlen(L, idx);
+	if (size > 1 || (is_field_present(L, "uri", idx) && size == 1) ||
+	    (!is_field_present(L, "uri", idx) && size == 0)) {
+		diag_set(IllegalParams, "Incorrect URI count in nested table");
+		return -1;
+	}
+	/* Table "default_params" is not allowed for single URI */
+	if (is_field_present(L, "default_params", idx)) {
+		diag_set(IllegalParams, "Default URI query parameters are "
+			 "not allowed for single URI");
+		return -1;
+	}
+	int rc = 0;
+	if (size == 1) {
+		lua_rawgeti(L, idx, 1);
+		rc = uri_create_from_string_or_number(uri, L);
+		lua_pop(L, 1);
+	} else {
+		lua_pushstring(L, "uri");
+		lua_rawget(L, idx);
+		assert(lua_type(L, -1) != LUA_TNIL);
+		rc = uri_create_from_string_or_number(uri, L);
+		lua_pop(L, 1);
+	}
+	if (rc != 0)
+		return rc;
+	return uri_add_params_from_lua_table(uri, L, "params", idx, true);
+}
+
+/**
+ * Create @a uri_set from the table, which located at the given valid @a idx.
+ */
+static int
+uri_set_create_from_lua_table(struct uri_set *uri_set, struct lua_State *L,
+			      int idx)
+{
+	const char *errmsg;
+	int rc = 0;
+	struct uri *uri = NULL;
+	assert(lua_type(L, idx) == LUA_TTABLE);
+	int uri_count = lua_objlen(L, idx);
+	/*
+	 * Since for replication empty table is allowed, so as not
+	 * to break backward compatibility we just return 0 here.
+	 */
+	if (uri_count == 0 && !is_field_present(L, "uri", idx))
+		return 0;
+	/*
+	 * If the number of numeric keys in the lua table is less
+	 * than or equal to one, it means that there is no more than
+	 * one URI in it.
+	 */
+	if (uri_count <= 1) {
+		uri = uri_set_add(uri_set, NULL);
```

Comment:
Such code needs additional copy of `struct uri`. Are you shure we want to do it?

Reply:
> Such code needs additional copy of struct uri. Are you shure we want to do it? I think about it, but implement this version to prevent additional copy of struct URI, which lead to many additional allocations

You can avoid it by switching to `uri_new` and `uri_delete`. Since a uri allocates a bunch of stuff internally, I think that would only be justified.

Reply:
I fixed in other way